ls1intum/Artemis
# `Development`: Fix flaky server test for student exams

### Checklist
#### General



### Motivation and Context



The testGradedFinalExamSummaryWithPlagiarismAndNotParticipatedBonusExamAsStudent test case is flaky means it can be failing and then be successful after restarting the test. This makes determining whether a change affects the functionality or not difficult, since the test needs to be restarted several times to know if the test is really failing or just flaky.

### Description

According to the logs when the test was failing [here](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS-JAVATEST/test/case/595540950), the test is always failing when the assigned student, as implemented [here](https://github.com/ls1intum/Artemis/blob/335c3d250f3d8058fd4168749d51062e590ae97a/src/test/java/de/tum/in/www1/artemis/exam/StudentExamIntegrationTest.java#L1937), is **studexamstudent2**. This is because the studexamstudent2 is already registered for the exam, as implemented [here](https://github.com/ls1intum/Artemis/blob/335c3d250f3d8058fd4168749d51062e590ae97a/src/test/java/de/tum/in/www1/artemis/exam/StudentExamIntegrationTest.java#L157). This means there could be 2 StudentExams with the same user, i.e. studexamstudent2. This makes [this API call](https://github.com/ls1intum/Artemis/blob/335c3d250f3d8058fd4168749d51062e590ae97a/src/test/java/de/tum/in/www1/artemis/exam/StudentExamIntegrationTest.java#L1956) in the test case would be failing, because studentExamRepository.findWithExercisesByUserIdAndExamId performed [here](https://github.com/ls1intum/Artemis/blob/335c3d250f3d8058fd4168749d51062e590ae97a/src/main/java/de/tum/in/www1/artemis/service/exam/ExamService.java#L340) will return non-unique result.

There is already an assertion on the test case [here](https://github.com/ls1intum/Artemis/blob/335c3d250f3d8058fd4168749d51062e590ae97a/src/test/java/de/tum/in/www1/artemis/exam/StudentExamIntegrationTest.java#L1952), but the exam id that is used there is different than what is used in the API, as implemented [here](https://github.com/ls1intum/Artemis/blob/335c3d250f3d8058fd4168749d51062e590ae97a/src/main/java/de/tum/in/www1/artemis/service/exam/ExamService.java#L340).

To fix the test case, 2 changes are implemented:
- Removing existing student exams whose student is the same as the assigned student to avoid having 2 StudentExams in the same Exam
- Fix [this assertion](https://github.com/ls1intum/Artemis/blob/335c3d250f3d8058fd4168749d51062e590ae97a/src/test/java/de/tum/in/www1/artemis/exam/StudentExamIntegrationTest.java#L1952) to use the correct exam id

### Steps for Testing
1. Run this shell script locally in the repo root folder:
```
mkdir log
for i in {1..50}
do
   echo "i=$i."
   ./gradlew cleanTest test --tests de.tum.in.www1.artemis.exam.StudentExamIntegrationTest -x npmInstall -x webapp 2> "log/$i.txt"
done
```
2. Verify that the test case **testGradedFinalExamSummaryWithPlagiarismAndNotParticipatedBonusExamAsStudent** is not failing by executing this command and no failed case is shown:
```
ls -f out | while read -r file; do cat out/$file | grep -i "StudentExamIntegrationTest > testGradedFinalExamSummaryWithPlagiarismAndNotParticipatedBonusExamAsStudent() FAILED"; done
```

### Review Progress






#### Code Review
#### Manual Tests

### Screenshots
<img width="1465" alt="image" src="https://user-images.githubusercontent.com/10885503/236193368-5d273e12-bd83-4e5c-9ba6-b1987f0c4c43.png">
